Implement artifact resolution in Docker and Buildpack builders

gsquared94 opened the pull request:
<!-- Thank you for your contribution! -->

<!-- Include if applicable: -->
**Related**: #4713, #4922
**Merge after**: #4893 

**Description** (Part 3 of several PRs.)
<!-- Describe your changes here. The more detail, the easier the review! -->
This PR implements resolving required artifacts as build args in docker builder; and as builder or run image in buildpacks, as described in the design for [supporting dependencies between build artifacts](https://github.com/GoogleContainerTools/skaffold/blob/master/docs/design_proposals/artifact-dependencies.md#docker-builder).
**User facing changes**
<!-- Describe any user facing changes this PR introduces. -->
<!-- "Before" and "After" sections work great - bonus points for screenshots! -->
<!-- Be sure all docs have been updated as well! -->
Modified example `integration/examples/microservices` app to test the required artifacts resolution as build args. Commands working:
```
skaffold dev --cache-artifacts=false --port-forward
```
```
skaffold build --cache-artifacts=false
```

#4893 will make this work with cache so it's a prereq PR. Will rebase and add tests after that gets in but general code review welcome!

tejal29 reviewed (commented):
One qq i had was adding `store` to the build signature. 
We have done this in past and there was a stage where the function signatures became crazy long. Every change affected multiple test files to add nil object to the test methods.
That is when we decided to introduce and add `runContext`. 
e.g. https://github.com/GoogleContainerTools/skaffold/pull/1824#pullrequestreview-215956698 and 
https://github.com/GoogleContainerTools/skaffold/pull/1885

I was wondering, if we could avoid the same here by  
e.g. have a method `WithArtifactStore` for builders that use the store.
I would also recommend breaking this PR in 2 stages
- Adding artifact resolution just to docker builder
- adding artifact resolution to buildpacks. 

Right now this PR is big.

I am not sure if `WithArtifactStore` would accomplish what you want to do. But i want to make sure we have investigated this direction where we avoid making changes to the interface.

gsquared94 commented:
> One qq i had was adding `store` to the build signature.
> We have done this in past and there was a stage where the function signatures became crazy long. Every change affected multiple test files to add nil object to the test methods.
> That is when we decided to introduce and add `runContext`.
> e.g. [#1824 (review)](https://github.com/GoogleContainerTools/skaffold/pull/1824#pullrequestreview-215956698) and
> #1885
> 
> I was wondering, if we could avoid the same here by
> e.g. have a method `WithArtifactStore` for builders that use the store.
> I would also recommend breaking this PR in 2 stages
> 
> * Adding artifact resolution just to docker builder
> * adding artifact resolution to buildpacks.
> 
> Right now this PR is big.
> 
> I am not sure if `WithArtifactStore` would accomplish what you want to do. But i want to make sure we have investigated this direction where we avoid making changes to the interface.

I also felt the parameter explosion happening to so many `Builder` interfaces. Do you think this interface with an added method is better than the one with an added parameter?
```go
type Builder interface {
	Build(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact) ([]Artifact, error)
	// Store returns an interface to update or get the build results
	Store() BuiltArtifacts
	// Prune removes images built with Skaffold
	Prune(context.Context, io.Writer) error
}
```

tejal29 commented:
> ```go
> Store() BuiltArtifacts
> ```

Thanks @gsquared94.  
The `Builder` should know how to
1. build artifacts
2. prune 

Would it be a builder's responsibility to update the built artifacts to a store? Should it know about a store?
I was reading on [Responsibility Driven Design](http://www.wirfs-brock.com/PDFs/A_Brief-Tour-of-RDD.pdf) and [0](https://www2.cs.arizona.edu/~mercer/Presentations/OOPD/12-RDD-Jukebox.pdf) and i feel like the builder should know just one think to build and clean up what it build. 

Embedding the `store` might mean we are giving too many responsibilities to it.
What do you think?

gsquared94 commented:
> > ```go
> > Store() BuiltArtifacts
> > ```
> 
> Thanks @gsquared94.
> The `Builder` should know how to
> 
> 1. build artifacts
> 2. prune
> 
> Would it be a builder's responsibility to update the built artifacts to a store? Should it know about a store?
> I was reading on [Responsibility Driven Design](http://www.wirfs-brock.com/PDFs/A_Brief-Tour-of-RDD.pdf), [0](https://www2.cs.arizona.edu/~mercer/Presentations/OOPD/12-RDD-Jukebox.pdf)
> I feel like the builder should know just one thing to build and clean up what it build.
> 
> Embedding the `store` might mean we are giving too many responsibilities to it.
> What do you think?

Good point. `store` is an implementation detail and probably the `Builder` implementation can encapsulate that. It's doesn't need to be in the interface method parameter list either.

gsquared94 commented:
closing this in favor of https://github.com/GoogleContainerTools/skaffold/pull/4932 where I've separated out the changes to only `docker builder`, rebased on master and changed the interface definitions according to @tejal29 's ideas.
